### ISSUE TITLE
[BUG] Inconsistency between Cellpose.eval() and CellposeModel.eval() and image input format

### ISSUE THREAD

**Issue by pakiessling** [MouseLand/cellpose#865]
cellpose version: 	3.0.1 
platform:       	linux 
python version: 	3.9.18 
torch version:  	2.0.0+cu117

I have an array of shape (7, 4914, 4914, 2) (z,x,y,c) that I want to segment every z-level separately with Cellpose.

The following works perfectly:
```
model = models.Cellpose(model_type="cyto3")
model2.eval(
    img,
    z_axis=0,
    channel_axis=3,
    diameter=250,
    flow_threshold=0.95,
    cellprob_threshold=-5,
    min_size=500,
    resample=False,
    do_3D=False,
    tile=True,
)
```

However when I load a model trained by me (in this casea fine-tuned model of cyto2):

<details open>
  <summary>Details of the model</summary>

  `{'torch': True, 'diam_mean': 30.0, 'mkldnn': True, 'device': device(type='cpu'), 'gpu': False, 'nchan': 2, 'nclasses': 3, 'nbase': [2, 32, 64, 128, 256], 'net': CPnet(
  (downsample): downsample(
    (down): Sequential(
      (res_down_0): resdown(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(2, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(2, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): Sequential(
            (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_2): Sequential(
            (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_3): Sequential(
            (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(2, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(2, 32, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_down_1): resdown(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): Sequential(
            (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_2): Sequential(
            (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_3): Sequential(
            (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(32, 64, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_down_2): resdown(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): Sequential(
            (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_2): Sequential(
            (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_3): Sequential(
            (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(64, 128, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_down_3): resdown(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): Sequential(
            (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_2): Sequential(
            (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_3): Sequential(
            (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(128, 256, kernel_size=(1, 1), stride=(1, 1))
        )
      )
    )
    (maxpool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (upsample): upsample(
    (upsampling): Upsample(scale_factor=2.0, mode='nearest')
    (up): Sequential(
      (res_up_0): resup(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(64, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=32, bias=True)
          )
          (conv_2): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=32, bias=True)
          )
          (conv_3): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=32, bias=True)
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(64, 32, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_up_1): resup(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(128, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=64, bias=True)
          )
          (conv_2): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=64, bias=True)
          )
          (conv_3): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(64, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=64, bias=True)
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(128, 64, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_up_2): resup(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(256, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=128, bias=True)
          )
          (conv_2): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=128, bias=True)
          )
          (conv_3): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(128, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=128, bias=True)
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(256, 128, kernel_size=(1, 1), stride=(1, 1))
        )
      )
      (res_up_3): resup(
        (conv): Sequential(
          (conv_0): Sequential(
            (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
            (1): ReLU(inplace=True)
            (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
          )
          (conv_1): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=256, bias=True)
          )
          (conv_2): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=256, bias=True)
          )
          (conv_3): batchconvstyle(
            (conv): Sequential(
              (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
              (1): ReLU(inplace=True)
              (2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
            )
            (full): Linear(in_features=256, out_features=256, bias=True)
          )
        )
        (proj): Sequential(
          (0): BatchNorm2d(256, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
          (1): Conv2d(256, 256, kernel_size=(1, 1), stride=(1, 1))
        )
      )
    )
  )
  (make_style): make_style(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (output): Sequential(
    (0): BatchNorm2d(32, eps=1e-05, momentum=0.05, affine=True, track_running_stats=True)
    (1): ReLU(inplace=True)
    (2): Conv2d(32, 3, kernel_size=(1, 1), stride=(1, 1))
  )
), 'pretrained_model': '/my_favorite_model', 'diam_labels': 95.8039, 'net_type': 'cellpose'}`
</details>

```
model = models.CellposeModel(
    pretrained_model="my_favorite_model"
)

model.eval(
    img,
    z_axis=0,
    channel_axis=3,
    diameter=250,
    flow_threshold=0.95,
    cellprob_threshold=-5,
    min_size=500,
    resample=False,
    do_3D=False,
    tile=True,
)

>ERROR: cannot process 4D images in 2D mode
>cellpose/transforms.py:494, in convert_image(x, channels, channel_axis, z_axis, do_3D, nchan)
>if not do_3D and x.ndim>3:
>   transforms_logger.critical('ERROR: cannot process 4D images in 2D mode')
>   raise ValueError('ERROR: cannot process 4D images in 2D mode')
```

Any idea what is going on?
How can I stop this transformation?

Thanks !

**Comment by pakiessling** [comment]
In addition I found difference between `nuclei `and `cyto3 `when segmenting (7, 4914, 4914, 2) (z,x,y,c)

```
img.shape
>(7, 4914, 4914, 2)

model = models.Cellpose(model_type="cyto3")

res = model.eval(
    img,
    z_axis=0,
    channel_axis=3,
    diameter=250,
    flow_threshold=0.95,
    cellprob_threshold=-5,
    min_size=500,
    resample=False,
    do_3D=False,
    tile=True,
)

res[0].shape
>(7, 4914, 4914)
------------------------------------------------------------------------------------------------------
model = models.Cellpose(model_type="nuclei")

img.shape
>(7, 4914, 4914, 1)

res = model2.eval(
    img,
    z_axis=0,
    channel_axis=3,
    diameter=55,
    flow_threshold=0.8,
    cellprob_threshold=-3,
    min_size=500,
    resample=False,
    do_3D=False,
    tile=True,
)

res[0].shape
>(4914, 4914)
```
This behavior also seems to have changed at some point as cellpose 1.0.2 (correctly?) returns (7, 4914, 4914) masks with the Nuclei model

**Comment by carsen-stringer** [comment]
I cannot reproduce your second bug @pakiessling (maybe I recently fixed that bug), but I've fixed the bug in which the two classes have different behavior -- basically printing a warning instead of an error

**Comment by pakiessling** [comment]
Thank you for the change @carsen-stringer.
I can indeed confirm that on `3.0.5.dev10+g2570475` both Cellpose.eval() and CellposeModel.eval() behave as expected.

However now Cellpose.eval() (but not CellposeModel.eval()) outputs an additonal warning: `channels not specified, using [0,0] (grayscale + no nuclei)`
Even when the channels should be unambiguous e.g:
```
img.shape
>(7, 4914, 4914, 2)

model = models.Cellpose(model_type="cyto3")

res = model.eval(
    img,
    z_axis=0,
    channel_axis=3,
    channels=[0, 1],
    diameter=250,
    flow_threshold=0.95,
    cellprob_threshold=-5,
    min_size=500,
    resample=False,
    do_3D=False,
    tile=True,
)
> channels not specified, using [0,0] (grayscale + no nuclei)
> 3D stack used, but stitch_threshold=0 and do_3D=False, so masks are made per plane only
```

**Comment by carsen-stringer** [comment]
Yes the user needs to specify the channels in the Cellpose case, so I set them to grayscale by default. in the CellposeModel case, the user might have a custom model with many channels, so they can actually set channels to None

**Comment by carsen-stringer** [comment]
hmm, I've removed the warning now (default is [0,0]), but channels = [0,1]
is equivalent to channels = [0,0] (no nuclei input). Do you mean [1,2] (you
want to use both channels)?

**Comment by pakiessling** [comment]
ah yes I wanted to use both available channels. I thought I had to specify them like that. Thank you!
